Support configuration via pyproject.toml

BrianPugh opened the pull request:
## Why is the change needed?
Addresses #182 .  Allows for creosote configuration via `pyproject.toml` under the `[tool.creosote]` table.

## What was done in this PR?
* `ARGS_CAN_BE_SPECIFIED_MULTIPLE_TIMES` was made into default behavior, **this means that this should be part of a `v3.0.0` release.** If we want to target a non-major release, we can revert this without too much effort (but I think v3.0.0 is the way to go).
* Each `add_argument` default was turned into a field default for the new `Config` dataclass. The diff looks big, but for each `add_argument`, only the `default` was removed. Its all added back at the end of the `parse_args` function via `parser.set_defaults(**dataclasses.asdict(defaults))`.
* On execution, `pyproject.toml` is loaded and `[tool.creosote]` is searched for. Provided parameters are loaded as CLI defaults. `pyproject.toml` keys have the same name as the `dest` variable name provided in argparse. toml file is loaded into new `Config` dataclass.  For example, the following is the relevant pyproject.toml section in one of my projects:

```toml
[tool.creosote]
venvs=[".venv"]
paths=["belay"]
deps-file="pyproject.toml"
sections=["tool.poetry.dependencies"]
exclude-deps =[
  "importlib_resources",
  "pydantic",
]
```

## Are there any concerns, side-effects, additional notes?
* README/Documentation will need to be updated.
* The feature is implemented with a simple dataclass and keyword-unpacking. If a more complicated, nested data structure is needed into the future, we should use some library (`cattrs`, `pydantic`, or `msgspec`), but the scope here is simple enough that those seem like overkill and this is simple enough. The scope of creosote also implies that this will be unnecessary.
* I didn't add any tests, tests would probably be nice.

## Checklist, when applicable

- [no] Added test(s)
- [no] Updated README.md
- [yes] Is this a backwards-compatibility breaking change?
- [yes] Resolves: #182

BrianPugh commented:
I *think* the remaining issues are related to python3.8 vs python3.7 features. I'd recommend dropping python 3.7 support since it reached EOL in June of 2023.

fredrikaverpil commented on src/creosote/cli.py:
```diff
@@ -1,19 +1,52 @@
+    sections: List[str] = field(default_factory=lambda: ["project.dependencies"])
+    exclude_deps: List[str] = field(default_factory=list)
+    deps_file: str = "pyproject.toml"
+    venvs: List[str] = field(default_factory=lambda: [".venv"])
```
I had requests in the past to auto-detect the currently activated virtual environment (#168). I was thinking the first value in this list could be `os.environ.get("VIRTUAL_ENV", ".venv")`.

```python
venvs: List[str] = field(default_factory=lambda: [os.environ.get("VIRTUAL_ENV", ".venv")])
```

Meaning; pick the path to the active virtual environment OR fall back to `".venv"`.

We can add this in as a separate change later, before the 3.0.0 release, but it should be part of the 3.0.0 release.

Do you foresee any issues with this? (I think it should be fine)
I suggest we leave a todo behind on this one:

```suggestion
    venvs: List[str] = field(default_factory=lambda: [".venv"])  # TODO: auto-detect active venv before releasing 3.0.0
```

BrianPugh replied:
addressed in 4a4435a4639c2bdc17a97

fredrikaverpil commented on src/creosote/cli.py:
```diff
@@ -71,73 +104,37 @@ def parse_args(args):
+        "--path",
+        dest="paths",
+        metavar="PATH",
+        action=CustomAppendAction,
```
```suggestion
        action=CustomAppendAction,  # TODO: add tests for v3-args before releasing 3.0.0
```

fredrikaverpil commented on src/creosote/cli.py:
```diff
@@ -71,73 +104,37 @@ def parse_args(args):
+        "--section",
+        dest="sections",
+        metavar="TOML_SECTION",
+        action=CustomAppendAction,
```
```suggestion
        action=CustomAppendAction,  # TODO: add tests for v3-args before releasing 3.0.0
```

fredrikaverpil commented on src/creosote/cli.py:
```diff
@@ -71,73 +104,37 @@ def parse_args(args):
+        "--exclude-dep",
+        dest="exclude_deps",
+        metavar="DEPENDENCY",
+        action="append",
```
```suggestion
        action="append",  # TODO: add tests for v3-args before releasing 3.0.0
```

fredrikaverpil commented on src/creosote/cli.py:
```diff
@@ -1,19 +1,52 @@
+    features: List[str] = field(default_factory=list)
+
+
+def load_defaults(src="pyproject.toml") -> Config:
+    """Load pyproject.toml defaults form user config.
+
+    Expects user configuration at ``[tool.creosote]``.
+    """
+    with open(src, "r", encoding="utf-8") as f:
+        project_config = toml.loads(f.read())
```
It looks like you always expect this file to be present in the current working directory. Please handle the case where `src` does not exist.
I believe creosote should just continue silently when the file does not exist, as this could be a common use case.

Please also add in debug logging for:
- file exists and is read.
- what gets read from the file's `[tool.creosote]` section.
- file does not exist.

BrianPugh replied:
added in 2e61193a5e7abd796

I had to refactor the argument parsing a bit so that logging-related configurations are set/parsed as early as possible.

BrianPugh replied:
whoops! definitely missed that.

* fixed missing file + added a unit test in 0f726c5065464bf7aa00e6.
* Added logging in 2e61193a5e7abd79

fredrikaverpil commented on src/creosote/cli.py:
```diff
@@ -1,19 +1,52 @@
+    features: List[str] = field(default_factory=list)
+
+
+def load_defaults(src="pyproject.toml") -> Config:
```
```suggestion
def load_defaults(src: str = "pyproject.toml") -> Config:
```

fredrikaverpil reviewed (changes requested):
I think this looks great! 😄👍 

It would be a good idea to add a test which verifies that the new `load_defaults` method works as expected. I would've added a test that creates [a temporary file](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html) which is then read into the function and asserts what happens when:
- pyproject.toml isn't found.
- pyproject.toml is found but toml section `[tool.creosote]` isn't found.
- the user has specified all possible key/value pairs in the toml.
- the user has specified a partial set of key/value pairs in the toml.

I also don't think I wrote any tests for the `ARGS_CAN_BE_SPECIFIED_MULTIPLE_TIMES` / `v3-args` feature, but that's something I can deal with outside of this PR.


EDIT: You can rebase, as I removed Python 3.7 in the main branch.

BrianPugh commented:
I think I addressed all the feedback, I left your TODO's in their about adding unit tests for the v3 argparsing; that can be added in another PR (this one is already getting a little big).

I also added a `__main__.py`, which allows people to invoke creosote as  `python -m creosote` (sometimes useful).

fredrikaverpil commented:
@BrianPugh you can run `ruff --fix src tests` to auto-fix the linting warnings, or manually address them of course.

fredrikaverpil commented:
@BrianPugh would you like me to push the related README changes to this PR or would you rather have me do that after merging this?

fredrikaverpil commented on tests/test_cli.py:
```diff
@@ -175,3 +174,22 @@ def test_unused_found_because_excluded_but_not_installed(  # noqa: PLR0913
+
+
+def test_load_defaults_no_file(tmp_path):
+    missing_pyproject = tmp_path / "pyproject.toml"
+    config = cli.load_defaults(missing_pyproject)
```
```suggestion
    empty_pyproject = tmp_path / "pyproject.toml"
    config = cli.load_defaults(empty_pyproject)
```

BrianPugh replied:
it's actually missing in this test case, however I just added more tests handling the other main scenarios.

BrianPugh replied:
addressed in 23a9f682c6fcd9c8ffd89e

fredrikaverpil commented on tests/test_cli.py:
```diff
@@ -175,3 +174,22 @@ def test_unused_found_because_excluded_but_not_installed(  # noqa: PLR0913
     assert exit_code == expected_exit_code
+
+
+def test_load_defaults_no_file(tmp_path):
+    missing_pyproject = tmp_path / "pyproject.toml"
+    config = cli.load_defaults(missing_pyproject)
+    assert config == cli.Config()  # Should return a default config
```
It would be great if you could complement this test with tests where the pyproject.toml is not empty, and:
- contains a full config and asserts on all the values.
- contains a partial config and asserts on all the values (of which some will use non-default and default values).

fredrikaverpil replied:
I see you added this in the PR description:

```toml
[tool.creosote]
venvs=[".venv"]
paths=["src"]
deps-file="pyproject.toml"
sections=["tool.poetry.dependencies"]
exclude-deps =[
  "importlib_resources",
  "pydantic",
]
```

It would be very nice if the assertions would show both the keys and the values, so we don't make any mistakes here. I'm thinking about the fact that there's a dash in `deps-file` for example, which I think is likely going to become an underscore.

BrianPugh replied:
addressed in 23a9f682c6fcd9c8ffd89e

fredrikaverpil commented on src/creosote/cli.py:
```diff
@@ -62,82 +101,33 @@ def parse_args(args):
+        action=CustomAppendAction,  # TODO: add tests for v3-args before releasing 3.0.0
+        help="pyproject.toml section(s) to scan for dependencies",
+    )
+    parser.add_argument(
+        "--exclude-dep",
```
For consistency;

```suggestion
    parser.add_argument(
        "-e",
        "--exclude-dep",
```

fredrikaverpil commented:
@BrianPugh do you also get this incomplete output when running `creosote --help`?

```bash
❯ creosote --help
usage: creosote [-h] [--verbose] [-f {default,no-color,porcelain}]

Prevent bloated virtual environments by identifing installed, but unused, dependencies

optional arguments:
  -h, --help            show this help message and exit
  --verbose             increase output verbosity (default: False)
  -f {default,no-color,porcelain}, --format {default,no-color,porcelain}
                        output format (default: None)
```

fredrikaverpil commented:
@BrianPugh 🤔 do you see this incomplete help too?

```bash
$ creosote --help
usage: creosote [-h] [--verbose] [-f {default,no-color,porcelain}]

Prevent bloated virtual environments by identifing installed, but unused, dependencies

optional arguments:
  -h, --help            show this help message and exit
  --verbose             increase output verbosity (default: False)
  -f {default,no-color,porcelain}, --format {default,no-color,porcelain}
                        output format (default: None)
```

BrianPugh commented:
Summary of changes since last review:
* Added more unit tests.
* Fixed linting.
* Fixed the broken help-string in [88eb81b](https://github.com/fredrikaverpil/creosote/pull/183/commits/88eb81bd8374d987108c5bf5692598837525ec7b).

fredrikaverpil commented on src/creosote/cli.py:
```diff
@@ -155,23 +144,49 @@ def parse_args(args):
 
 def main(args_=None):
-    args = parse_args(args_)
+    init_parser = argparse.ArgumentParser(
+        add_help=False,
+    )
+    # Immediately parse verbosity
+    init_parser.add_argument(
+        "--verbose",
+        dest="verbose",
+        action="store_true",
+        help="increase output verbosity",
+    )
+    init_parser.add_argument(
+        "-f",
+        "--format",
+        dest="format",
+        choices=typing.get_args(Config.__annotations__["format"]),
+        help="output format",
+    )
+    args, _ = init_parser.parse_known_args(args_)
 
+    # Configure logger early
     formatters.configure_logger(verbose=args.verbose, format_=args.format)
 
+    # Now that the logger is configured, configure the rest of the parser.
+    parser = argparse.ArgumentParser(
+        description=(
+            "Prevent bloated virtual environments by identifing installed, "
+            "but unused, dependencies"
+        ),
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        parents=[init_parser],
+    )
+    configure_parser(parser, defaults=load_defaults())
+    args = parser.parse_args(args_)
```
Hmmm, okay so this means you can't configure the logging as part of `pyproject.toml`. And this is only because I asked you about logging of the `pyproject.toml` file loading?

This all seems kind of messy to me. Let me think about this. Seems like we could load the `pyproject.toml` later, instead of configuring parts of arguments and logging/formatting early.

fredrikaverpil replied:
I pushed a second commit, so to simplify this all a bit. Rather than logging in the `load_defaults` function, I just make sure it can return valuable information and I log this a bit later.

fredrikaverpil reviewed (approved):
Awesome, thank you! 🎉
